src/tasks/AotCompilerTask/MonoAOTCompiler.cs

```diff
@@ -357,47 +357,77 @@ private bool ExecuteInternal()
 
         _cache = new FileCache(CacheFilePath, Log);
 
-        //FIXME: check the nothing changed at all case
+        List<PrecompileArguments> argsList = new();
+        foreach (var assemblyItem in _assembliesToCompile)
+            argsList.Add(GetPrecompileArgumentsFor(assemblyItem, monoPaths));
 
         _totalNumAssemblies = _assembliesToCompile.Count;
-        int allowedParallelism = Math.Min(_assembliesToCompile.Count, Environment.ProcessorCount);
-        if (BuildEngine is IBuildEngine9 be9)
-            allowedParallelism = be9.RequestCores(allowedParallelism);
-
-        if (DisableParallelAot || allowedParallelism == 1)
+        if (CheckAllUpToDate(argsList))
         {
-            foreach (var assemblyItem in _assembliesToCompile)
-            {
-                if (!PrecompileLibrarySerial(assemblyItem, monoPaths))
-                    return !Log.HasLoggedErrors;
-            }
+            Log.LogMessage(MessageImportance.High, "Everything is up-to-date, nothing to precompile");
+
+            _fileWrites.AddRange(argsList.SelectMany(args => args.ProxyFiles).Select(pf => pf.TargetFile));
+            foreach (var args in argsList)
+                compiledAssemblies.GetOrAdd(args.AOTAssembly.ItemSpec, args.AOTAssembly);
         }
         else
         {
-            ParallelLoopResult result = Parallel.ForEach(
-                                                    _assembliesToCompile,
-                                                    new ParallelOptions { MaxDegreeOfParallelism = allowedParallelism },
-                                                    (assemblyItem, state) => PrecompileLibraryParallel(assemblyItem, monoPaths, state));
+            int allowedParallelism = Math.Min(_assembliesToCompile.Count, Environment.ProcessorCount);
+            if (BuildEngine is IBuildEngine9 be9)
+                allowedParallelism = be9.RequestCores(allowedParallelism);
 
-            if (!result.IsCompleted)
+            if (DisableParallelAot || allowedParallelism == 1)
             {
-                return false;
+                foreach (var args in argsList)
+                {
+                    if (!PrecompileLibrarySerial(args))
+                        return !Log.HasLoggedErrors;
+                }
             }
+            else
+            {
+                ParallelLoopResult result = Parallel.ForEach(
+                                                        argsList,
+                                                        new ParallelOptions { MaxDegreeOfParallelism = allowedParallelism },
+                                                        (args, state) => PrecompileLibraryParallel(args, state));
```

Comment:
The part also missing here is cancellation. Do you need to implement `ICancelableTask` and pass in a `CancellationToken`?

I think if you started a build and hit Ctrl+C (or the Cancel button in VS), the build would hang until the `Parallel.ForEach()` completed.

Reply:
I have opened a separate issue for this - https://github.com/dotnet/runtime/issues/59211 . I'll implement this for the MonoAOTCompiler, and EmccCompile task.